sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/Offset.scala

```diff
@@ -17,11 +17,14 @@
 
 package org.apache.spark.sql.execution.streaming
 
+import com.fasterxml.jackson.annotation.JsonTypeInfo
+
 /**
  * An offset is a monotonically increasing metric used to track progress in the computation of a
  * stream. An [[Offset]] must be comparable, and the result of `compareTo` must be consistent
  * with `equals` and `hashcode`.
  */
+@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
```

Comment:
In my experience (with jackson specifically and people who have read the docs) you end up with surprises down the road due to magic.

@zsxwing In this particular case, maybe we should remove the offsets from the status messages, or turn the to strings or something.  These are going to be implemented by specific sinks, so I'm not sure we want to expose the details of arbitrary classes to the listener bus.